### PR TITLE
TFT: Allow to sleep after 5mn (backlight)

### PR DESCRIPTION
**Pull request by tpruvot**
Prepare public methods for remote or centralised wakeup/sleep...

Add TOUCH_IDLE_SLEEP in seconds to enable the feature

Will also blank dogm screen on sleep (CLASSIC_UI) and do the same with COLOR_UI

Tested on STM32 & STM32F1 HALs (should work on LPC1768 too)

### CODE REVIEW

File: Marlin/src/lcd/dogm/u8g_dev_tft_upscale_from_128x64.cpp [hunk 1]
```diff
@@ -340,6 +347,17 @@ static uint8_t page;
...
 #endif // HAS_TOUCH_BUTTONS
 
+static void u8g_upscale_clearScreen(u8g_t *u8g, u8g_dev_t *dev) {
+  setWindow(u8g, dev, 0, 0, (TFT_WIDTH) - 1, (TFT_HEIGHT) - 1);
+  #if HAS_LCD_IO
+    tftio.WriteMultiple(TFT_MARLINBG_COLOR, (TFT_WIDTH) * (TFT_HEIGHT));
+  #else
+    memset2(buffer, TFT_MARLINBG_COLOR, (TFT_WIDTH) / 2);
+    for (uint16_t i = 0; i < (TFT_HEIGHT) * sq(GRAPHICAL_TFT_UPSCALE); i++)
+      u8g_WriteSequence(u8g, dev, (TFT_WIDTH) / 2, (uint8_t *)buffer);
```
Comment (tpruvot) [inline Marlin/src/lcd/dogm/u8g_dev_tft_upscale_from_128x64.cpp:356]: argh, found a problem here without HAS_LCD_IO... sigh

File: Marlin/src/lcd/tft/ui_320x240.cpp [hunk 2]
```diff
@@ -45,8 +45,24 @@
   #include "../../feature/bedlevel/bedlevel.h"
 #endif
 
+#if HAS_TOUCH_SLEEP
+  static bool sleepCleared;
```
Comment (tpruvot) [inline Marlin/src/lcd/tft/ui_320x240.cpp]: may be moved in tft_idle function, if static

**Comment by tpruvot** [comment]
To note: this feature will need a better implementation to allow buttons or encoder to wake up the display (for those who have one, like mks) Maybe some generic "idle/power" class could be added somewhere... and maybe/optionally skip the tft sleep during prints. Also, im working on the backlight brightness with the PWM... different behaviors could be wanted... backlight dimming on idle instead of a "blank" (black) screen...

**Comment by tpruvot** [comment]
Ok, the feature seems correct for most common cases now... maybe parts could be simplified/refactored like the completion_feedback function... If someone has a touchscreen with a rotary encoder, could be nice to give a try... (edit: found one)

**Comment by tpruvot** [comment]
rebased and partially squashed...

**Comment by tpruvot** [comment]
So for the encoder, it doesnt wake up on rotation for now, just on rotary button press (and custom buttons)... will need more changes, now i can play with the lerdge tft. (**edit:** fixed) I may create a special class to handle the idle delay and just call wakeup on touchscreen click like i did for the feedback... was not proper to put that in touchscreen class directly. **Edit:** moved to shared classes Touch (color_ui)/TouchButtons (classic_ui)

File: Marlin/src/HAL/STM32/tft/xpt2046.cpp [hunk 3]
```diff
@@ -33,6 +33,7 @@
 uint16_t delta(uint16_t a, uint16_t b) { return a > b ? a - b : b - a; }
 
 SPI_HandleTypeDef XPT2046::SPIx;
+millis_t XPT2046::lastTouch;
```
Comment (thinkyhead) [inline Marlin/src/HAL/STM32/tft/xpt2046.cpp]: I suggest that all these added things be wrapped in `#if TOUCH_IDLE_SLEEP > 0` so they are never compiled unless the option is enabled.

Comment (tpruvot) [reply Marlin/src/HAL/STM32/tft/xpt2046.cpp]: yep, i will try to move code from xpt class to Marlin/src/lcd/tft/touch.cpp, idle() func already check time... but its only for COLOR UI... so in TouchButtons class too (its one level higher but should work for all touchscreens then...)

**Comment by thinkyhead** [comment]
I've moved `HAS_TOUCH_SLEEP` to `Conditionals_LCD.h` and moved some of the TFT common code over to `ui_common.cpp`. Made a few other cosmetic changes and everything looks good to me at this point.

**Pull request merged**